File: content/snippets/typescript/interview_lcci/16/18_pattern_matching.ts
function patternMatching(pattern: string, value: string): boolean {
    if (pattern.length === 0) return value.length === 0;

    let countA = 0;
    let countB = 0;
    for (const ch of pattern) {
        if (ch === "a") countA += 1;
        else countB += 1;
    }

    if (countA < countB) {
        const swapped = [...pattern].map((ch) => (ch === "a" ? "b" : "a")).join("");
        return patternMatching(swapped, value);
    }

    if (value.length === 0) return countB === 0;

    const n = value.length;
    for (let lenA = 0; countA * lenA <= n; lenA++) {
        const rest = n - countA * lenA;
        let lenB = 0;
        if (countB === 0) {
            if (rest !== 0) continue;
        } else {
            if (rest % countB !== 0) continue;
            lenB = rest / countB;
        }

        let pos = 0;
        let a: string | null = null;
        let b: string | null = null;
        let ok = true;
        for (const ch of pattern) {
            if (ch === "a") {
                const sub = value.slice(pos, pos + lenA);
                if (a === null) a = sub;
                else if (a !== sub) {
                    ok = false;
                    break;
                }
                pos += lenA;
            } else {
                const sub = value.slice(pos, pos + lenB);
                if (b === null) b = sub;
                else if (b !== sub) {
                    ok = false;
                    break;
                }
                pos += lenB;
            }
        }
        if (ok && a !== b) return true;
    }
    return false;
}

console.log(patternMatching("abba", "dogcatcatdog"));
console.log(patternMatching("abba", "dogcatcatfish"));

File: content/snippets/python/interview_lcci/16/18_pattern_matching.py
def pattern_matching(pattern: str, value: str) -> bool:
    if not pattern:
        return value == ""

    count_a = pattern.count("a")
    count_b = pattern.count("b")

    if count_a < count_b:
        swapped = "".join("b" if c == "a" else "a" for c in pattern)
        return pattern_matching(swapped, value)

    if value == "":
        return count_b == 0

    n = len(value)
    for len_a in range(0, n // count_a + 1):
        rest = n - count_a * len_a
        if count_b == 0:
            if rest != 0:
                continue
            len_b = 0
        else:
            if rest % count_b != 0:
                continue
            len_b = rest // count_b

        pos = 0
        a = None
        b = None
        ok = True
        for ch in pattern:
            if ch == "a":
                sub = value[pos:pos + len_a]
                if a is None:
                    a = sub
                elif a != sub:
                    ok = False
                    break
                pos += len_a
            else:
                sub = value[pos:pos + len_b]
                if b is None:
                    b = sub
                elif b != sub:
                    ok = False
                    break
                pos += len_b
        if ok and a != b:
            return True
    return False


def main() -> None:
    print(pattern_matching("abba", "dogcatcatdog"))
    print(pattern_matching("abba", "dogcatcatfish"))


if __name__ == "__main__":
    main()
